darts/preprocessing/pipeline.py

```diff
@@ -0,0 +1,248 @@
+"""
+Pipeline
+--------
+"""
+from copy import deepcopy
+from collections import defaultdict
+from typing import List, Optional, Union, Tuple, Iterator
+
+from darts.logging import raise_if_not, raise_if, get_logger
+from darts import TimeSeries
+from darts.preprocessing.base_transformer import BaseTransformer
+
+logger = get_logger(__name__)
+
+
+class Pipeline:
+    def __init__(self,
+                 transformers: List[BaseTransformer[TimeSeries]],
+                 names: Optional[List[str]] = None,
+                 deep: bool = False):
+        """
+        Pipeline combines multiple transformers chaining them together.
+
+        Parameters
+        ----------
+        transformers
+            List of transformers.
+        names
+            Optional list of names for transformers must have same length as transformers and names should be unique.
+            If no names are specified transformer name will be set to str value of his index in original list.
+        deep
+            If set prevents changes in original transformers made by pipeline.
+        """
+
+        raise_if_not(all((isinstance(t, BaseTransformer)) for t in transformers),
+                     "transformers should be objects deriving from BaseTransformer", logger)
+        if deep:
+            self._transformers = deepcopy(transformers)
+        else:
+            self._transformers = transformers
+
+        if self._transformers is None or len(self._transformers) == 0:
+            self._transformers: List[BaseTransformer[TimeSeries]] = []
+            logger.warning("Empty pipeline created")
+
+        if names:
+            raise_if_not(len(names) == len(self._transformers), "names should have same length as transformer", logger)
+            raise_if_not(len(names) == len(set(names)), "names should be unique", logger)
+        else:
+            names = list(map(str, range(len(self._transformers))))
+
+        self._name_2_idx = {
+            str(name): idx for idx, name in enumerate(names)
+        }
+        self._names = names
+
+        self._args_for_transformers = defaultdict(tuple)
+        self._kwargs_for_transformers = defaultdict(dict)
+
+        self._reversible = all((t.reversible for t in self._transformers))
+        self._can_predict = len(self._transformers) and self._transformers[-1].can_predict
+
+    def set_transformer_args_kwargs(self, key: Union[str, int], *args, **kwargs):
```

Comment:
Do you mean similar function but on transformer? That would be simper, will add it here. Do you want this to be removed?

Reply:
yes, I wonder why do we need setters altogether - Transformers could take init() parameters to constructors and then, after creation be immutable - no need to set things inside (unless Im missing some point here)

Reply:
One remark because of fit, how @grll mentioned in the previous discussion, transformers won't be immutable. I like your approach it is much cleaner. correct me if I am wrong, then transform and other functions would not need any additional parameters and I think for most use cases that would be fine. This would simplify a lot in implementation but one thing that sticks to me is that we then cannot modify behaviour of transformer. If transformer does something based on some alpha hyper parameter then to experiment with different alpha values we would need to create new transformer and new pipeline with that transformer.

Reply:
I am not sure if i understand correctly but i think we can leverage `Transformer`'s mutability if we want to change some of its parameters. In such case we can simply store this transformer as variable before passing to `Pipeline` and if needed we can change its parameters using created variable instead of `set_transfomer_args_kwargs` method of `Pipeline`. It will probably mean moving this method to `Transformer` class but still I think it is more convenient than doing it in pipeline (we could then get rid of `names` argument if I am correct).  What do you think?